collector/lib/CollectorConfig.cpp

```diff
@@ -438,17 +453,88 @@ void CollectorConfig::HandleConfig(const std::filesystem::path& filePath) {
-                << ". Error: " << e.what();
+    std::string errorMsg = "Failed to open the configuration file: " + filePath.string() + ". Error: " + e.what();
+    LogRuntimeConfigError(errorMsg);
   } catch (const YAML::ParserException& e) {
-    CLOG(FATAL) << "Failed to parse the configuration file: " << filePath
-                << ". Error: " << e.what();
+    std::string errorMsg = "Failed to parse the configuration file: " + filePath.string() + ". Error: " + e.what();
+    LogRuntimeConfigError(errorMsg);
   } catch (const YAML::Exception& e) {
-    CLOG(FATAL) << "An error occurred while loading the configuration file: " << filePath
-                << ". Error: " << e.what();
+    std::string errorMsg = "An error occurred while loading the configuration file: " + filePath.string() + ". Error: " + e.what();
+    LogRuntimeConfigError(errorMsg);
   } catch (const std::exception& e) {
-    CLOG(FATAL) << "An unexpected error occurred while trying to read: " << filePath << e.what();
+    std::string errorMsg = "An unexpected error occurred while trying to read: " + filePath.string() + e.what();
+    LogRuntimeConfigError(errorMsg);
+  }
+}
+
+void WaitForFileToExist(const std::filesystem::path& filePath) {
+  while (!std::filesystem::exists(filePath)) {
+    sleep(1);
+  }
+}
+
+void WaitForInotifyAddWatch(int& fd, int& wd, const std::filesystem::path& filePath) {
+  while (true) {
+    fd = inotify_init();
+    if (fd < 0) {
+      CLOG_THROTTLED(ERROR, std::chrono::seconds(30)) << "inotify_init failed";
+      continue;
+    }
+    int wd = inotify_add_watch(fd, filePath.c_str(), IN_MODIFY | IN_MOVE_SELF | IN_DELETE_SELF);
+    if (wd < 0) {
+      CLOG_THROTTLED(ERROR, std::chrono::seconds(30)) << "Failed to add inotify watch for " << filePath;
+      close(fd);
+    } else {
+      return;
+    }
+    sleep(1);
+  }
+}
+
+void CollectorConfig::WatchConfigFile(const std::filesystem::path& filePath) {
+  CLOG(INFO) << "In WatchConfigFile";
+  int fd, wd;
+  WaitForFileToExist(filePath);
+  WaitForInotifyAddWatch(fd, wd, filePath);
+  HandleConfig(filePath);
+
+  char buffer[1024];
+  while (true) {
+    CLOG(INFO) << "Waiting for file event";
+    int length = read(fd, buffer, sizeof(buffer));
+    if (length < 0) {
+      CLOG(ERROR) << "Unable to read event for " << filePath;
+    }
+
+    for (int i = 0; i < length; i += sizeof(struct inotify_event)) {
+      struct inotify_event* event = (struct inotify_event*)&buffer[i];
+      CLOG(INFO) << "Got event event->mask= " << event->mask;
+      CLOG(INFO) << "event->mask & IN_MODIFY= " << (event->mask & IN_MODIFY);
+      if (event->mask & IN_MODIFY) {
+        CLOG(INFO) << "File was replaced";
+        HandleConfig(filePath);
+      } else if ((event->mask & IN_MOVE_SELF) || (event->mask & IN_DELETE_SELF) || (event->mask & IN_IGNORED)) {
+        CLOG(INFO) << "File was removed";
+        WaitForFileToExist(filePath);
+        WaitForInotifyAddWatch(fd, wd, filePath);
```

Comment:
Aren't you leaking `wd` and `fd` here?

Reply:
This was intentional so that fd and wd would be available for
```
  inotify_rm_watch(fd, wd);
  close(fd);
```
but with the changes that you suggest, this is not needed.